presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java

```diff
@@ -164,4 +182,373 @@ private void assertTableLayout(Plan plan, String tableName, TupleDomain<Subfield
         assertEquals(layoutHandle.getDomainPredicate(), domainPredicate);
         assertEquals(layoutHandle.getRemainingPredicate(), remainingPredicate);
     }
+
+    @Test
+    public void testPushdownArraySubscripts()
+    {
+        assertUpdate("CREATE TABLE test_pushdown_array_subscripts(id bigint, a array(bigint), b array(array(varchar)))");
+
+        assertPushdownSubscripts("test_pushdown_array_subscripts");
+
+        // Unnest
+        assertPushdownSubfields("SELECT t.b, a[1] FROM test_pushdown_array_subscripts CROSS JOIN UNNEST(b) as t(b)", "test_pushdown_array_subscripts",
+                ImmutableMap.of("a", toSubfields("a[1]")));
+
+        assertPushdownSubfields("SELECT t.b, a[1] FROM test_pushdown_array_subscripts CROSS JOIN UNNEST(b[1]) as t(b)", "test_pushdown_array_subscripts",
+                ImmutableMap.of("a", toSubfields("a[1]"), "b", toSubfields("b[1]")));
+
+        assertPushdownSubfields("SELECT t.b[2], a[1] FROM test_pushdown_array_subscripts CROSS JOIN UNNEST(b) as t(b)", "test_pushdown_array_subscripts",
+                ImmutableMap.of("a", toSubfields("a[1]"), "b", toSubfields("b[*][2]")));
+
+        assertPushdownSubfields("SELECT id, grouping(index), sum(length(b[1][2])) FROM test_pushdown_array_subscripts CROSS JOIN UNNEST(a) as t(index) GROUP BY grouping sets ((index, id), (index))", "test_pushdown_array_subscripts",
+                ImmutableMap.of("b", toSubfields("b[1][2]")));
+
+        assertPushdownSubfields("SELECT id, b[1] FROM test_pushdown_array_subscripts CROSS JOIN UNNEST(a) as t(unused)", "test_pushdown_array_subscripts",
+                ImmutableMap.of("b", toSubfields("b[1]")));
+
+        // No subfield pruning
+        assertPushdownSubfields("SELECT array_sort(a)[1] FROM test_pushdown_array_subscripts", "test_pushdown_array_subscripts",
+                ImmutableMap.of());
+
+        assertPushdownSubfields("SELECT id FROM test_pushdown_array_subscripts CROSS JOIN UNNEST(a) as t(index) WHERE a[1] > 10 AND cardinality(b[index]) = 2", "test_pushdown_array_subscripts",
+                ImmutableMap.of());
+
+        assertUpdate("DROP TABLE test_pushdown_array_subscripts");
```

Comment:
Is this safe in a multithreaded context?  Should we mark the test as single threaded?

Reply:
I think so. Each test is using different table name.